pkg/services/secrets/kvstore/migrations/migrator.go

```diff
@@ -9,41 +9,55 @@ import (
 	"github.com/grafana/grafana/pkg/infra/serverlock"
 	datasources "github.com/grafana/grafana/pkg/services/datasources/service"
 	"github.com/grafana/grafana/pkg/services/secrets/kvstore"
+	"github.com/grafana/grafana/pkg/setting"
 )
 
 var logger = log.New("secret.migration")
 
+const actionName = "secret migration task "
+
 // SecretMigrationService is used to migrate legacy secrets to new unified secrets.
 type SecretMigrationService interface {
 	Migrate(ctx context.Context) error
 }
 
 type SecretMigrationServiceImpl struct {
-	Services          []SecretMigrationService
-	ServerLockService *serverlock.ServerLockService
+	services                 []SecretMigrationService
+	ServerLockService        *serverlock.ServerLockService
+	migrateToPluginService   *kvstore.MigrateToPluginService
+	migrateFromPluginService *kvstore.MigrateFromPluginService
 }
 
 func ProvideSecretMigrationService(
+	cfg *setting.Cfg,
 	serverLockService *serverlock.ServerLockService,
 	dataSourceSecretMigrationService *datasources.DataSourceSecretMigrationService,
-	pluginSecretMigrationService *kvstore.PluginSecretMigrationService,
+	migrateToPluginService *kvstore.MigrateToPluginService,
+	migrateFromPluginService *kvstore.MigrateFromPluginService,
 ) *SecretMigrationServiceImpl {
 	services := make([]SecretMigrationService, 0)
 	services = append(services, dataSourceSecretMigrationService)
-	// pluginMigrationService should always be the last one
-	services = append(services, pluginSecretMigrationService)
+	// plugin migration should always be last; should either migrate to or from, not both
+	if cfg.SectionWithEnvOverrides("secrets").Key("migrate_from_plugin").MustBool(false) {
```

Comment:
I think we talked about this, but as far as I can see, this will mean that if we have a plugin, we either do the migration from or to the plugin, there is no way to skip the migration